### PR TITLE
Add file verification to GET calls for ftp/s and sftp

### PR DESCRIPTION
**Pull request by johngoza**

### CODE REVIEW

File: lib/sftp.js [hunk 1]
```diff
@@ -34,18 +68,37 @@ const getFileViaSftp =
...
+
+const verifyAndGetFileViaSftp = def("verifyAndGetFileViaSftp")
+({})
+([SshClientType, SftpConnectionConfig, FutureType($.String)(ReadStreamType)])
+(client => connectionConfig =>
+  setupConnection(client)(connectionConfig) // todo: magic constant is not okay but needs to be included here until next pr
+    .pipe(chain (client => verifyFile(client)(0)(3)(connectionConfig)("")))
+    .pipe(chain (getFileViaSftp(client))));
 
 const sendFileViaSftp =
```
Comment (mreinigjr) [inline lib/sftp.js:98]: Here is my suggestion for this function

```const sendFileViaSftp =
    def("sendFileViaSftp")
    ({})
    ([SshClientType, ReadStreamType, SftpConnectionConfig, FutureType($.String)($.String)])
    (client => readStream => connectionConfig => Future((reject, resolve) => {
      client.on("error", (err) => {
        client.end();
        reject(err);
      });

      client.on("ready", () => {
        client.sftp((err, sftp) => {
          if (err) {
            client.end();
            reject(err);
          } else {
            const writeStream = sftp.createWriteStream(connectionConfig.remoteFileName);
            writeStream.on("error", (err) => {
              client.end();
              reject(err);
            });

            writeStream.on("ready", () => {
              readStream.pipe(writeStream);
              resolve("Upload successful");
            });

            writeStream.on("close", () => {
              client.end();
            });
          }
        });
      });

      // todo: probably here too
      client.connect(connectionConfig);
      return () => { };
    }));```

File: lib/sftp.js [hunk 2]
```diff
@@ -76,12 +129,16 @@ const sendFileViaSftp =
         });
       });
 
+      // todo: probably here too
```
Comment (mreinigjr) [inline lib/sftp.js:132]: what does this `todo` mean?

File: tests/unit/lib/ftp-tests.js [hunk 3]
```diff
@@ -96,48 +109,126 @@ describe("Unit Tests - ftp.js", function() {
...
+    //
+    //   fork
+    //   (err => {
+    //     expect(err).to.equal("503 get failed");
+    //     done();
+    //   })
+    //   (done)
+    //   (getFileViaFtp(mockFtpClient)(mockConnectionConfig));
+    // });
+  });
```
Comment (mreinigjr) [inline tests/unit/lib/ftp-tests.js:145]: Should this be deleted?

File: tests/unit/lib/ftp-tests.js [hunk 4]
```diff
@@ -96,48 +109,126 @@ describe("Unit Tests - ftp.js", function() {
...
+      mockFtpClient.connect = () => { };
+      mockFtpClient.get = () => { };
+      mockFtpClient.on = () => { };
+      mockFtpClient.put = () => { };
       mockFtpClient.end = () => { };
 
       fork
       (err => {
-        expect(err).to.equal("503 get failed");
+        expect(err).to.deep.equal("400 Directory not found");
```
Comment (mreinigjr) [inline tests/unit/lib/ftp-tests.js:214]:
```suggestion
        expect(err).to.equal("400 Directory not found");
```

File: tests/unit/lib/ftp-tests.js [hunk 5]
```diff
@@ -238,7 +336,265 @@ describe("Unit Tests - ftp.js", function() {
...
+        callback(null, readable);
+      };
+      mockFtpClient.list = (remoteDirectory, cb) => {
+        cb(null, mockFileList);
+      };
+      mockFtpClient.end = () => { };
+
+      fork
+      (done)
+      (data => {
```
Comment (mreinigjr) [inline tests/unit/lib/ftp-tests.js:447]: you could have just checked that the result was an instance of a readable.  Either that, you should update the description of your test.

File: tests/unit/lib/ftp-tests.js [hunk 6]
```diff
@@ -238,7 +336,265 @@ describe("Unit Tests - ftp.js", function() {
...
+        callback(mockError);
+      };
+      mockFtpClient.list = (remoteDirectory, cb) => {
+        cb(null, mockFileList);
+      };
+      mockFtpClient.end = () => { };
+
+      fork
+      (err => {
+        expect(err).to.deep.equal("500 an error occurred");
```
Comment (mreinigjr) [inline tests/unit/lib/ftp-tests.js:502]:
```suggestion
        expect(err).to.equal("500 an error occurred");
```

File: tests/unit/lib/ftp-tests.js [hunk 7]
```diff
@@ -238,7 +336,265 @@ describe("Unit Tests - ftp.js", function() {
...
+      (done)
+      (data => {
+        let result = "";
+
+        data.on("data", function(d) {
+          result += d.toString();
+        });
+
+        data.on("end", function() {
+          expect(result).to.deep.equal("hello world");
```
Comment (mreinigjr) [inline tests/unit/lib/ftp-tests.js:455]:
```suggestion
          expect(result).to.equal("hello world");
```

File: tests/unit/lib/ftp-tests.js [hunk 8]
```diff
@@ -238,7 +336,265 @@ describe("Unit Tests - ftp.js", function() {
...
+      (err => {
+        expect(err).to.deep.equal(expectedError);
+        done();
+      })
+      (done)
+      (setupConnection(mockFtpClient)(mockConnectionConfig));
+    });
+  });
+
+  describe("verifyAndGetFileViaFtp", function() {
```
Comment (mreinigjr) [inline tests/unit/lib/ftp-tests.js:408]: `verifyAndGetFileViaFtp` is composition of several functions that are already unit tested.  This particular suite of tests should be in the integration folder.

File: tests/unit/lib/ftp-tests.js [hunk 9]
```diff
@@ -238,7 +336,265 @@ describe("Unit Tests - ftp.js", function() {
...
+        accessCount > 0 ? cb(null, mockFileList1) : cb(null, mockFileList2);
+        accessCount = accessCount + 1;
+      };
+      mockFtpClient.end = () => { };
+
+      const expectedError = "File metadata changed while attempting GET. File is not currently viable for consumption";
+
+      fork
+      (err => {
+        expect(err).to.deep.equal(expectedError);
```
Comment (mreinigjr) [inline tests/unit/lib/ftp-tests.js:593]:
```suggestion
        expect(err).to.equal(expectedError);
```

File: tests/unit/lib/sftp-tests.js [hunk 10]
```diff
@@ -130,16 +138,126 @@ describe("Unit Tests - sftp.js", function() {
...
+      mockSftpClient.sftp = (cb) => {
+        cb(null, sftp);
+      };
+      mockSftpClient.connect = () => { };
+      mockSftpClient.end = () => { };
+
+      fork
+      (done)
+      (data => {
+        expect(data).to.deep.equal(mockFileList);
```
Comment (mreinigjr) [inline tests/unit/lib/sftp-tests.js:179]: `expect(data).to.deep.equal(mockFileList);` 
:+1:

`deep` should be used on objects that you have.  Sometimes it is funky with arrays though.

File: tests/unit/lib/sftp-tests.js [hunk 11]
```diff
@@ -130,16 +138,126 @@ describe("Unit Tests - sftp.js", function() {
...
+      const mockSftpClient = new EventEmitter();
+      mockSftpClient.sftp = (cb) => {
+        cb(null, sftp);
+      };
+      mockSftpClient.connect = () => { };
+      mockSftpClient.end = () => { };
+
+      fork
+      (err => {
+        expect(err).to.deep.equal("400 Directory not found");
```
Comment (mreinigjr) [inline tests/unit/lib/sftp-tests.js:246]:
```suggestion
        expect(err).to.equal("400 Directory not found");
```

File: tests/unit/lib/sftp-tests.js [hunk 12]
```diff
@@ -255,7 +375,332 @@ describe("Unit Tests - sftp.js", function() {
...
+      mockSftpClient.connect = () => {
+        mockSftpClient.emit("ready");
+      };
+      mockSftpClient.on("ready", () => {});
+      mockSftpClient.end = () => {};
+
+      fork
+      (done)
+      (data => {
+        expect(data).to.deep.equal(mockSftpClient);
```
Comment (mreinigjr) [inline tests/unit/lib/sftp-tests.js:405]: :+1:

File: tests/unit/lib/sftp-tests.js [hunk 13]
```diff
@@ -255,7 +375,332 @@ describe("Unit Tests - sftp.js", function() {
...
+      (err => {
+        expect(err).to.deep.equal(mockError);
+        done();
+      })
+      (done)
+      (setupConnection(mockSftpClient)(mockConnectionConfig));
+    });
+  });
+
+  describe("verifyAndGetFileViaSftp", function() {
```
Comment (mreinigjr) [inline tests/unit/lib/sftp-tests.js:445]: should also live in integration test folder

File: tests/unit/lib/sftp-tests.js [hunk 14]
```diff
@@ -255,7 +375,332 @@ describe("Unit Tests - sftp.js", function() {
...
+      })
+      (done)
+      (verifyAndGetFileViaSftp(mockSftpClient)(mockConnectionConfig));
+
+      readable.emit("error", mockError);
+      readable.push(null);
+    });
+  });
+
+  describe("verifyFile", function() {
```
Comment (mreinigjr) [inline tests/unit/lib/sftp-tests.js:563]: this too should be in integration test folder

File: tests/unit/lib/utility-functions-tests.js [hunk 15]
```diff
@@ -49,198 +71,268 @@ describe("Unit Tests - UtilityFunctions", function() {
     });
   });
 
-  describe("validateConnectionConfig", function() {
-    it("should resolve with the config if all values are valid", function(done) {
-      const config = {
-        "host": "some-site",
-        "port": 2,
-        "remoteFilePath": "some-path",
+  describe("filterFileMetadata", function() {
```
Comment (mreinigjr) [inline tests/unit/lib/utility-functions-tests.js:74]: Is `filterFileMetadata` really reusable across all 3 channels right now?

File: tests/unit/lib/utility-functions-tests.js [hunk 16]
```diff
@@ -49,198 +71,268 @@ describe("Unit Tests - UtilityFunctions", function() {
...
-      (done)
-      (validateConnectionConfig(config));
+      const result = validateConnectionConfig(config);
+      expect(result).to.deep.equal(S.Left(expectedError));
     });
 
-    it("should reject if host is null", function(done) {
+    it("should reject if host is null", function() {
       const config = {
         "host": "",
```
Comment (mreinigjr) [inline tests/unit/lib/utility-functions-tests.js:185]: `host` is not `null`, but rather an empty string.

So since you are using `R.isNull` and `R.isEmpty`, you will need to have tests for both the empty string and for when it is `null`.

File: tests/unit/lib/utility-functions-tests.js [hunk 17]
```diff
@@ -49,198 +71,268 @@ describe("Unit Tests - UtilityFunctions", function() {
...
-        done();
-      })
-      (done)
-      (validateConnectionConfig(config));
+      const result = validateConnectionConfig(config);
+      expect(result).to.deep.equal(S.Left(expectedError));
     });
 
-    it("should reject if remoteFilePath is null", function(done) {
+    it("should reject if remoteFileName is null", function() {
```
Comment (mreinigjr) [inline tests/unit/lib/utility-functions-tests.js:272]: again, you are testing the empty string case and not the null case.

**Review by mreinigjr — CHANGES_REQUESTED** [review]
In the dependency requirement statements, please list the full file name.  Just doing `.../ftp` might get us in trouble vs `.../ftp.js`.

We need to start making these PRs smaller if possible.

File: tests/unit/lib/ftp-tests.js (hunk 4 above)
Comment (johngoza) [reply tests/unit/lib/ftp-tests.js:214]: deep equal is silly for simple data types; use for nested only. :me: === :dummy:

File: tests/unit/lib/utility-functions-tests.js (hunk 17 above)
Comment (johngoza) [reply tests/unit/lib/utility-functions-tests.js:285]: image

Comment (johngoza) [reply tests/unit/lib/utility-functions-tests.js:285]: not sure why github is being bad here

File: tests/unit/lib/utility-functions-tests.js (hunk 16 above)
Comment (johngoza) [reply tests/unit/lib/utility-functions-tests.js:198]: again unsure why github is being bad here

**Review by mreinigjr — APPROVED** [review]
